pkg/services/query/query.go

```diff
@@ -90,18 +90,41 @@ func (s *Service) QueryData(ctx context.Context, user *user.SignedInUser, skipCa
 	g, ctx := errgroup.WithContext(ctx)
 	results := make([]backend.Responses, len(byDataSource))
 
+	// Create panic recovery function for loop below
+	recoveryFn := func(results *[]backend.Responses, queries []*simplejson.Json) {
+		if r := recover(); r != nil {
+			var err error
+			s.log.Error("query datasource panic", "error", r, "stack", log.Stack(1))
+			if theErr, ok := r.(error); ok {
+				err = theErr
+			} else if theErrString, ok := r.(string); ok {
+				err = fmt.Errorf(theErrString)
+			} else {
+				err = fmt.Errorf("unexpected error, see the server log for details")
+			}
+			// Because of the panic, there is no valid response for any query. Append an error for each one.
+			*results = append(*results, buildErrorResponses(err, queries))
+		}
+	}
+
+	// Query each datasource concurrently
 	for _, queries := range byDataSource {
 		rawQueries := make([]*simplejson.Json, len(queries))
 		for i := 0; i < len(queries); i++ {
 			rawQueries[i] = queries[i].rawQuery
 		}
 		g.Go(func() error {
 			subDTO := reqDTO.CloneWithQueries(rawQueries)
+			// Handle panics in the datasource qery
+			defer recoveryFn(&results, subDTO.Queries)
 
 			subResp, err := s.QueryData(ctx, user, skipCache, subDTO)
 
 			if err == nil {
 				results = append(results, subResp.Responses)
+			} else {
+				// Because of the error, there is no valid response for any query. Append the error for each one.
+				results = append(results, buildErrorResponses(err, subDTO.Queries))
```

Comment:
I'm pretty sure this code (and above) is not goroutine safe, i.e. multiple goroutines could write to the `results` at the same time and crash.